Add support for File System multiple instances

ramakrishnapallala opened the pull request:
Added support for having multiple instances of a file system in Zephyr.

subsys: fs: Extend storage_dev type beyond 'struct device'
ext: fs: fat: Enable multiple volume support
subsys: disk: Add support for multiple disk interfaces
subsys: fs: Add the support for multiple instances of fs
subsys: usb: Add the support for mass storage disk name
samples: subsys: usb: Set disk name Kconfig option
tests: subsys: fs: Set FAT FS mount point name as per ELM FAT
tests: subsys: fs: Add test application for FAT FS dual instance

Commit:
subsys: fs: Extend storage_dev type beyond 'struct device'

Extend storage_dev type beyond 'struct device' by changing
the variable type to void pointer. This will be needed if we
have additional transactional layers b/w filesystem and the
actual backend device.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

tyrantlizard77 reviewed (commented):
Can you give an example of these intermediate objects?

We have some "drivers" which are actually intermediate subsystems built on base driver types. These too are struct device and have defined subsystems.

I'd like to know what the bigger picture you are going for here is.

ramakrishnapallala commented:
@andrewboie at this stage, this is primarily needed for FatFS where we handle things differently based on the type of storage (Flash, RAM or SDHC). So making storage type opaque helps us abstract the low-level details.

ramakrishnapallala commented:
@nvlsianpu @carlescufi 

After doing a thorough analysis of NFFS library code and having tried few ideas to enable multiple instances of NFFS, I had come to the conclusion that it is not possible to have multiple instances of NFFS with current library code.

The reason is, NFFS is implemented as singleton file system which uses multiple flash areas to store information in the form of files and directories. Ideally, we can see these individuals or group of flash areas as logical partitions which can be formatted and mounted as NFFS file system. But the NFFS implementation does not provide any means to group these flash areas as a logical partition or drive and does not allow us to format or mount multiple NFFS instances.

The following global variables defines the NFFS file system and which can only form a single instamce of a file system:
struct nffs_area nffs_areas[NFFS_CONFIG_MAX_AREAS];
uint8_t nffs_num_areas;
uint8_t nffs_scratch_area_idx;
uint16_t nffs_block_max_data_sz;
struct nffs_area_desc *nffs_current_area_descs;
struct nffs_inode_entry *nffs_root_dir;
struct nffs_inode_entry *nffs_lost_found_dir;

Let me know if you think otherwise and share your inputs.

nvlsianpu commented on include/fs.h:
```diff
@@ -58,7 +58,7 @@ struct fs_mount_t {
 	const char *mnt_point;
 	void *fs_data;
-	struct device *storage_dev;
+	void *storage_dev;
```
I think it might be helpfully if this will be a pointer to a union. This will make the future code more slef-documenting than the pointer to void. What do you thin?

ramakrishnapallala replied:
my intention of changing this to void* is, if we need have more info other than device then the application can package that in a struct and pass it to the FS. For example, in NFFS case I thought of sending the no.of Flash areas and device pointer for the specific drive/partition.

nvlsianpu replied:
ok - understood this form the beginning. It might be not good idea to reference variety of sub-types to the union I proposed.
<b>Accepted</b> as is.

nvlsianpu commented:
@ramakrishnapallala 
Of Course you are right - for now NFFS code is able to support only one NFFS instance. I didn't check the scope of global variables (what you listed looks reasonable) - I can  look at this scope essentially if you need this.

ramakrishnapallala commented:
@nvlsianpu thanks for the confirmation on NFFS multiple instances.

Commit:
subsys: fs: Add the support for multiple instances of fs

Add support for multiple instances of a file system by
making use of mount point as the disk volume name which
is used by the file system library while formatting or
mounting a disk.

Also moved out file system specific data structures from
public fs.h header and handled them in corresponding
file system interface files by introducing open files and
open directories concept which is already being used in
NFFS interface module. Now it is extended to FatFs as well.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

Commit:
samples: subsys: usb: Set disk name Kconfig option

Assign the appropriate disk name Kconfig option to the
usb mass storage sample test.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

ramakrishnapallala commented:
@nvlsianpu @carlescufi @lpereira @andrewboie Please review the PR and let me know your feedback.

lpereira commented on ext/fs/fat/zfs_diskio.c:
```diff
@@ -29,13 +29,15 @@
 DSTATUS disk_status(BYTE pdrv)
 {
-	if (disk_access_status() != 0) {
+	if (disk_access_status(pdrv_str[pdrv]) != 0) {
```
Please add an assertion here to ensure that pdrv is within the bounds of pdrv_str.  While it'll be removed in release mode, it provides some guidance to perform validations in syscalls later.

ramakrishnapallala replied:
Ok

lpereira commented on subsys/disk/disk_access.c:
```diff
@@ -0,0 +1,179 @@
+#include <logging/sys_log.h>
+
+/* list of mounted file systems */
+static sys_dlist_t disk_access_list;
```
Please use SYS_DLIST_STATIC_INIT

ramakrishnapallala replied:
I prefer dynamic as I want to start from disk_access_init I do not want to initialize the before hand.
do you see any issue in not using static init?

lpereira commented on subsys/disk/disk_access.c:
```diff
@@ -0,0 +1,179 @@
+static sys_dlist_t disk_access_list;
+
+/* lock to protect storage layer registration */
+static struct k_mutex mutex;
```
Please use K_SEM_DEFINE

ramakrishnapallala replied:
same.

lpereira commented on subsys/disk/disk_access.c:
```diff
@@ -0,0 +1,179 @@
+	return rc;
+}
+
+static int disk_init(struct device *dev)
```
If the mutex and the list are initalized statically, you can get rid of this function

ramakrishnapallala replied:
I would like to still have disk_init as placeholder for future purpose.

lpereira commented on subsys/disk/disk_access.c:
```diff
@@ -0,0 +1,179 @@
+		 * shorter than longest_match match or if path
+		 * name is shorter than the mount point name.
+		 */
+		if (strlen(itr->name) != name_len) {
```
Do we know if itr->name will always point to a 0-terminated string?

ramakrishnapallala replied:
the names are constant strings.

lpereira commented on subsys/disk/disk_access.c:
```diff
@@ -0,0 +1,179 @@
+		}
+
+		/* Check for disk name match */
+		if (strncmp(name, itr->name, name_len) == 0) {
```
Will this be an issue if we have two disks with the same prefix?

ramakrishnapallala replied:
we can not have two disk with same name

nvlsianpu replied:
?

lpereira commented on subsys/disk/disk_access_flash.c:
```diff
@@ -274,3 +275,25 @@ int disk_access_ioctl(u8_t cmd, void *buff)
 	return -EINVAL;
 }
+
+static struct disk_operations flash_disk_ops = {
```
Please make this const

lpereira commented on subsys/disk/disk_access_flash.c:
```diff
@@ -274,3 +275,25 @@ int disk_access_ioctl(u8_t cmd, void *buff)
+	.ioctl = disk_flash_access_ioctl,
+};
+
+static struct disk_info flash_disk = {
```
Please make this const if possible

lpereira commented on subsys/disk/disk_access_ram.c:
```diff
@@ -77,3 +81,25 @@ int disk_access_ioctl(u8_t cmd, void *buff)
 	return 0;
 }
+
+static struct disk_operations ram_disk_ops = {
```
Please make this const

lpereira commented on subsys/disk/disk_access_ram.c:
```diff
@@ -77,3 +81,25 @@ int disk_access_ioctl(u8_t cmd, void *buff)
+	.ioctl = disk_ram_access_ioctl,
+};
+
+static struct disk_info ram_disk = {
```
Please make this const if possible

lpereira reviewed (commented):
Didn't finish reviewing the patch, but I have some comments that I'd like to get out of the way.  Nothing serious.

nvlsianpu commented:
I will review this In a few days

ramakrishnapallala commented:
@nvlsianpu can you have a look at the PR?

ramakrishnapallala commented:
@lpereira I have addressed your comments. Please have a look. Thanks.

@nvlsianpu Please review the PR as it is destined for v1.12 release.

carlescufi commented:
CC @lemrey

nvlsianpu commented:
@ramakrishnapallala I'm reviewing this now.

nvlsianpu commented:
2'nd commit `ext: fs: fat: Enable multiple volume support` and 3'rd commit `subsys: disk: Add support for multiple disk interfaces` should be merged as changes are closely linked.

Commit:
subsys: disk: Add support for multiple disk interfaces

Add support for enabling multiple disk interfaces (Flash, RAM)
simultaneously in Zephyr by introducing a simple disk interface
framework where we can register multiple disks which would
interface with different storage devices. This would enable us
to have multiple instances of FATFS in Zephyr.

Add support for mass storage drive disk name which will be
used as an argument when calling the disk interface API's.

Enable multiple volumes support configuration in
ELM FAT library.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

ramakrishnapallala commented on subsys/disk/Kconfig:
```diff
@@ -13,9 +13,28 @@ config DISK_ACCESS
 	help
 	  Enable disk access over a supported media backend like FLASH or RAM
 
+config DISK_ACCESS_MAX_VOLUMES
```
If we enable FatFS then this should be same as ffconf.h otherwise it doesn't matter

ramakrishnapallala commented:
@nvlsianpu I have addressed your comment and updated the PR.

One additional change included is, I have added open file and directory Kconfigs at subsys/fs level. I have introduced the open files concept to fatfs as well which NFFS is already doing it.

Commit:
tests: subsys: fs: Add test for FAT FS dual instance case

Add a test application to demonstrate the dual instances of
FAT file system.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>

ramakrishnapallala commented:
@nvlsianpu Please have a look at the update the PR. Thanks.

ramakrishnapallala commented on subsys/fs/Kconfig:
```diff
@@ -49,6 +49,16 @@ config FILE_SYSTEM_NFFS
 	  Note: NFFS requires 1-byte unaligned access to flash thus it
 	  will not work on devices that support only aligned flash access.
 
+config FILE_SYSTEM_NUM_FILES
```
Ok, will add separate Kconfig for now. we need to see how we can unify the configs in future.

Commit:
tests: subsys: fs: Add changes to support multiple FS instance

Set FAT FS mount point name as per ELM FAT Library drive
name format.

Add file system specific headers wherever necessary.

Signed-off-by: Ramakrishna Pallala <ramakrishna.pallala@intel.com>
